File: chapter14/cpp/chapter_14.cpp
#include "chapter_14.h"


std::ostream &operator << (std::ostream &out, const std::vector<int> &vect )
{
  std::cout << "[";
  for (auto ele: vect)
  {
    out << ele << ',';
  }
  std::cout << "]\n";
  return out;
}

std::ostream &operator << (std::ostream &out, const std::vector<std::pair<std::string, int> > &vect )
{
  std::pair<std::string, int> ele;
  std::cout << '[';
  for (int i = 0; i < vect.size(); i++)
  {
    ele = vect[i];
    out << "key:" << ele.first << ',' << "value:" << ele.second << '|';
  }
  std::cout << "]\n";
  return out;
}

void HashMap::display()
{
  for (int i = 0; i < table.size() ; i++)
  {
    std::cout << table[i];
  }
}


int HashMap::hashCode(std::string str)
{
  int charCode = 0;
  int hashCode = 0;

  if (str.size() == 0)
  {
    return charCode;
  }

  for (int i = 0; i < str.size() ; i++)
  {
    charCode = (unsigned) str[i];
    hashCode += charCode;
  }
  return hashCode;
}

void HashMap::add(std::string key, int value)
{
  int index = hashCode(key) % capacity;
  std::vector<std::pair<std::string, int> > &collection = table[index];
  std::pair<std::string, int> pair = std::make_pair(key, value);;

  if (collection.empty())
  {
    collection.push_back(pair);
  }
  else
  {
    // search for key/value
    for (int i = 0; i < collection.size(); i++)
    {
      if (collection[i].first == key  && collection[i].second == value)
      {
        return;
      }
    }
    // add key/value
    collection.push_back(pair);
  }
  size++;
}

bool HashMap::empty()
{
  return size == 0;
}

std::vector<int> HashMap::find(std::string key)
{
  int index = hashCode(key) % capacity;
  std::vector<std::pair<std::string, int> > &collection = table[index];
  std::vector<int> result;

  for (std::pair<std::string, int>  pairData: collection )
  {
    if (pairData.first == key)
    {
      result.push_back(pairData.second);
    }
  }
  return result;
}

int HashMap::remove(std::string key)
{
  int index = hashCode(key) % capacity;
  std::vector<std::pair<std::string, int> > &collection = table[index];
  int result = 0xDEADBEEF;
  int pos = 0;

  for (std::pair<std::string, int> &pairData: collection )
  {
    if (pairData.first == key)
    {
      result = pairData.second;
      break;
    }
    pos++;
  }

  if (pos < collection.size())
  {
    size--;
    collection.erase(collection.begin() + pos);
  }

  return result;
}

double HashMap::load_factor()
{
  return size / capacity;
}

void HashMap::grow ()
{
  int increase = capacity - capacity + int(capacity * 0.50);
  int bucket_index;
  int code;
  std::pair<std::string, int> pair;

  // update capacity
  capacity = capacity + increase;

  // increase array size
  for (; increase--; )
  {
    table.push_back(std::vector<std::pair<std::string, int> >());
  }

  // rehash
  for (int i = 0; i < capacity; i++)
  {
    bucket_index = 0;
    while (bucket_index < table[i].size())
    {
      pair = table[i][bucket_index];
      code = hashCode(pair.first);

      if (code % capacity != i )
      // delete and add
      {
        table[i].erase(table[i].begin() + bucket_index);
        add(pair.first, pair.second);
      }
      else
      {
        bucket_index++;
      }
    }
  }

}

void HashMap::set_size (const int &newCap)
{
  std::pair<std::string, int> pair;
  int bucket_index;
  int count;
  int code;

  if (newCap <= 0)
  {
    return;
  }

  if (newCap > capacity)
  {
    count = newCap - capacity;
    for (int i = 0; i < count; i++)
    {
      table.push_back(std::vector<std::pair<std::string, int> >());
    }
  }
  else
  {
    table = std::vector<std::vector <std::pair<std::string, int> > >(table.begin(), table.begin() + newCap);
  }

  capacity = newCap;

  // rehash
  for (int i = 0; i < capacity; i++)
  {
    bucket_index = 0;
    while (bucket_index < table[i].size())
    {
      pair = table[i][bucket_index];
      code = hashCode(pair.first);
      if (code % capacity != i )
      {
        table[i].erase(table[i].begin() + bucket_index);
        add(pair.first, pair.second);
      }
      else
      {
        bucket_index++;
      }
    }
  }
}

void HashMap::addMap( HashMap &obj)
{
  int index;
  int j = -1;

  for (int i = 0; i < obj.table.size(); i++)
  {
    std::vector<std::pair<std::string, int> > &obj_collection = obj.table[i];
    for (std::pair<std::string, int> pair: obj_collection)
    {
      index = hashCode( pair.first) % capacity;

      // check THIS BUCKET for duplicate key
      for ( j = 0; j < table[index].size(); j++)
      {
        if (table[index][j].first == pair.first)
        {
          table[index][j].second = pair.second;
          break;
        }
      }

      if (j >= table[index].size())
      {
        table[index].push_back(pair);
      }
    }
  }
}

void HashMap::select_keys( std::vector<std::pair<std::string, int> > keys )
{
  std::pair<std::string, int> pair;
  int j, k;

  // iterate buckets
  for (int i = 0; i < table.size(); i++)
  {
    // bucket - iterate elements
    j = 0;
    while ( j < table[i].size() )
    {
      pair = table[i][j];
      for (k = 0; k < keys.size(); k++)
      {
        if (keys[k].first == pair.first ) // equal keys found
        {
          break;
        }
      }

      if (k >= keys.size() ) // pair key not found in keys list
      {
        table[i].erase(table[i].begin() + j);
      }
      else
      {
        ++j;
      }
    }
  }
}

void HashMultiMap::add(std::string key, int value)
{
  int index = this->hashCode(key) % capacity;
  std::vector<std::pair<std::string, int> > &collection = table[index];
  std::pair<std::string, int> pair = std::make_pair(key, value);;

  if (collection.empty())
  {
    collection.push_back(pair);
  }
  else
  {
    collection.push_back(pair);
  }
  size++;
}


int main()
{
  HashMultiMap hasher;
  hasher.add("hector", 21);
  hasher.add("hector",300);

  hasher.display();
  // hasher.grow();
  // std::cout << '\n';
  // hasher.set_size(11);
  // hasher.display();

  auto x = hasher.find("hector");
}

File: chapter14/java/chapter14.java
package algo;
import algo.Chapter6.*;
import java.util.ArrayList;

public class Chapter14 {
	
	public static class Hash 
	{
		private int capacity;
		private ArrayList< ArrayList<String> > table;
		
		Hash(int cap) 
		{
			if (cap <= 0)
			{
				cap = 10;
			}
			
			capacity = cap;
			table = new ArrayList< ArrayList<String>  >();
			
			for (int i = 0; i < cap; i++)
			{
				table.add(null);
			}
		}
		
		private int hashCoder (String key) 
		{
			int hash = 0;
			
			for (int i = 0; i < key.length(); i++ )
			{
				hash += key.codePointAt(i);
			}
			return hash;
		}
		
		public void add (String key, int value)
		{
			int code = hashCoder(key);
			int index = code % capacity;
			int i = 0;
			
			if ( table.get(index) == null) 
			{
				table.set(index, new ArrayList<String>() );
			}
			
			// check for value
			for (i = 0; i <  table.get(index).size(); i+=2)
			{
				if (table.get(index).get(i).compareTo(key) == 0 && table.get(index).get(i + 1).compareTo(String.valueOf(value)) == 0)
				{
					return;
				}
			}
			
			// add new elements to array 
			 table.get(index).add(key);
			 table.get(index).add(String.valueOf(value));

			
		}
		
		public boolean isEmpty() 
		{
			boolean empty = true;
			for (ArrayList<String> retArr: table)
			{
				if (retArr != null) // skip nulled array elements 
				{
					empty &= retArr.size() == 0;
				}
			}
			return empty;
		}

		public ArrayList<Integer> find(String key)
		{
			int code = hashCoder(key);
			int index = code % capacity;
			ArrayList<Integer> buffer = new ArrayList<Integer>();
			int[] result = {};
			
			if (table.get(index) == null) 
			{
				return buffer;
			}
			
			
			for (int i = 0; i < table.get(index).size(); i+=2)
			{
				if (table.get(index).get(i).compareTo(key) == 0)
				{
					buffer.add( Integer.valueOf(table.get(index).get(i + 1) ));
				}
			}
			
			return buffer;
		}
		
		public void displayHashTree () // DEBUG
		{
			for (ArrayList<String > retStringArr: table )
			{
				if (retStringArr != null) 
				{
					System.out.print("\t[");
					for (int i = 0 ; i < retStringArr.size(); i++) 
					{
						System.out.print(retStringArr.get(i) + ",");
					}
					System.out.print("]");
					System.out.println();
				}
				else 
				{
					System.out.println("\t" + retStringArr);					
				}
			}
		}
		
		public ArrayList<Integer> remove (String key) 
		{
			int index = hashCoder(key) % capacity;
			ArrayList<Integer> buffer = new ArrayList<Integer>();
			int i = 0;
			
			if (table.get(index) != null) 
			{
				while (i < table.get(index).size())
				{
					if ( table.get(index).get(i).compareTo(key) == 0   )
					{
						// add value
						buffer.add( Integer.valueOf( table.get(index).get(i + 1) ));
						
						// delete 
						table.get(index).remove(i);
						table.get(index).remove(i);
					}
					else 
					{
						i+=2;
					}
				}
			}
			
			return buffer;
		}
		
		public double loadFactor () 
		{
			int elementCount = 0;
			int bucketCount = 0;
			
			for (ArrayList<String> arr : table) 
			{
				if (arr != null) 
				{
					bucketCount += 1;
					elementCount += arr.size() / 2;
				}
			}
			
			if (bucketCount == 0) 
			{
				return -1;
			}
			
			return elementCount / (double) bucketCount;
		}
		
		public void setSize (int newCap)
		{
			ArrayList< ArrayList<String> > newTable = new ArrayList<ArrayList<String>>();
			String value; 
			String key;
			int index;
			
			capacity = newCap; 
			
			for (int i = 0; i < capacity; i ++)
			{
				newTable.add(null);
			}
			
			// get keyValue pairs 
			for (ArrayList<String> arr : table) 
			{
				if (arr != null) 
				{
					for (int i = 0; i < arr.size(); i+=2) 
					{
						key = arr.get(i);
						value = arr.get(i + 1);
						index = hashCoder(key) % capacity;
						
						if (newTable.get(index) == null) 
						{
							newTable.set(index, new ArrayList<String>());
						}
						
						newTable.get(index).add(key);
						newTable.get(index).add(value);
					}
				}
			}
			
			// update class instance variable table 
			table = newTable;
		}
		
		public void grow ()
		{
			setSize(capacity + capacity / 2);
		}
		
		public void addMap (Hash hashmap)
		{
			String value; 
			String key;
			int index;
			boolean foundKeyValuePair;
			
			// get keyValue pairs from hashmap argument 
			for (ArrayList<String> arr : hashmap.table) 
			{
				if (arr != null) 
				{
					for (int i = 0; i < arr.size(); i+=2) 
					{
						key = arr.get(i);
						value = arr.get(i + 1);
						index = hashCoder(key) % capacity;
						
						// search for key/value pair in 'this' table
						foundKeyValuePair = false;
						for (int k = 0 ; k < table.get(index).size() ; k+=2 )
						{
							foundKeyValuePair |= table.get(index).get(i).compareTo(key) == 0 && table.get(index).get(i + 1).compareTo(value) == 0;
						}
						
						// add to table (if necessary)
						if (!foundKeyValuePair)
						{
							table.get(index).add(key);
							table.get(index).add(value);
						}
						
					}
				}
			}
		}
		
		public void selectKeys (String[] keys) 
		{
			ArrayList<String> replacementBucket;
	
			for (int i = 0; i < table.size(); i++) 
			{
				if (table.get(i) != null)
				{
					
					replacementBucket = new ArrayList<String>();
					
					for (int k = 0; k < table.get(i).size(); k += 2)
					{
						//  compare to keys list 
						for (String selectedKey: keys)
						{
							if ( selectedKey.compareTo(table.get(i).get(k)) == 0)
							{
								// selection (add to new bucket) 
								replacementBucket.add(table.get(i).get(k));
								replacementBucket.add(table.get(i).get(k + 1));
							}
						}
					}
					
					// replace bucket 
					table.set(i, replacementBucket);
				}
			}
		}
		
	}
}
